lib/autoload/polyglot/language.rb

```diff
@@ -0,0 +1,90 @@
+# An abstract language. This is an ActiveRecord class for other models to reference to a family of
+# languages.
+#
+# Do *NOT* remove languages after they have been defined because a database record corresponds to
+# a class implemented here.
+#
+# Each language acts as a singleton class, call +Polyglot::Language.instance+ to retrieve the
+# instance for the language.
+class Polyglot::Language < ActiveRecord::Base
+  extend ActiveSupport::Autoload
+
+  self.table_name = 'polyglot_languages'
+  acts_as_forest
+
+  # Eager load all languages defined in the language directory.
+  Dir["#{__dir__}/language/*"].each { |language| require language }
+
+  after_initialize :set_readonly
+  after_save :set_readonly
+
+  validate :unique_root_language, unless: :parent
+
+  # @!method self.instance
+  #   Gets an instance of the language object. Any class which implements this method can be
+  #   instantiated and is expected to no longer be abstract.
+  #
+  #   @abstract
+  #   @return [Polyglot::Language]
+
+  # Gets the display name of the language.
+  #
+  # @abstract
+  # @return [String]
+  def self.display_name
+    fail NotImplementedError
+  end
+
+  # The stylesheets that need to be packaged with the rest of the application.
+  #
+  # This should include the Rouge/Pygments stylesheet for formatting code.
+  #
+  # @abstract
+  # @return [Array<String>]
+  def self.stylesheets
+    fail NotImplementedError
+  end
+
+  # The script files that need to be packaged with the rest of the application.
+  #
+  # This should include the Ace mode for the language.
+  #
+  # @abstract
+  # @return [Array<String>]
+  def self.javascript
+    fail NotImplementedError
+  end
+
+  # Finds or creates the root instance for languages of this class.
+  #
+  # @return [Polyglot::Language]
+  def self.root
+    # Creating the root record would call +root+, so if we are defining @root, we return nil.
+    if instance_variable_defined?(:@root)
+      @root
+    else
+      @root = nil
+      @root = find_or_create_by(name: display_name, parent: nil)
+    end
+  end
+  private_class_method :root
+
+  # Loads or creates the database records for each defined programming language.
+  def self.load_languages
```

Comment:
no no, the singleton is in each of the language classes. I am both creating the record in the database and loading the record for each class.

Remember each language has a root -- that is what is in the singleton variable of each class. That is in `instance` which is what I'm forcing the evaluation of here